### PR TITLE
[Method References 2] LPS-165735 Convert method reference to ServiceTrackerMap in modules commerce-discount-service and commerce-service

### PR DESCRIPTION
**Pull request by W2jx**

### CODE REVIEW

File: modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/price/CommerceProductPriceCalculationV2Impl.java [hunk 1]
```diff
@@ -395,30 +395,19 @@ public CommerceMoney getUnitPrice(
...
-
-		String commercePriceListDiscoveryKey = GetterUtil.getString(
-			properties.get("commerce.price.list.discovery.key"));
-
-		_commercePriceListDiscoveryMap.remove(commercePriceListDiscoveryKey);
+	@Activate
+	protected void activate(BundleContext bundleContext) {
+		_serviceTrackerMap = ServiceTrackerMapFactory.openSingleValueMap(
+			bundleContext, CommercePriceListDiscovery.class, null,
+			(serviceReference, emitter) -> emitter.emit(
```
Comment (dantewang) [inline modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/price/CommerceProductPriceCalculationV2Impl.java:402]: @W2jx You should be able to use `com.liferay.osgi.service.tracker.collections.map.ServiceTrackerMapFactory#openSingleValueMap(org.osgi.framework.BundleContext, java.lang.Class<S>, java.lang.String)` directly in this pull, if you want to keep the usage of component property and the services are provided by Liferay (provider type).

**Review by dantewang — CHANGES_REQUESTED** [review]

**Comment by catarineaguiar** [comment]
ci:test:relevant

**Comment by catarineaguiar** [comment]
ci:stop

**Comment by catarineaguiar** [comment]
no regressions found.

**Comment by W2jx** [comment]
ci:test:relevant

**Comment by W2jx** [comment]
Hi @dantewang , I have used `com.liferay.osgi.service.tracker.collections.map.ServiceTrackerMapFactory#openSingleValueMap(org.osgi.framework.BundleContext, java.lang.Class<S>, java.lang.String) ` directly.

**Comment by zhangyan0701** [comment]
No regressions found.

**Comment by W2jx** [comment]
no conflicts(1/5 8:55)

**Comment by W2jx** [comment]
no conflicts(1/9 16:16)

File: modules/apps/commerce/commerce-discount-api/src/main/java/com/liferay/commerce/discount/application/strategy/CommerceDiscountApplicationStrategyRegistry.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ */
+
+package com.liferay.commerce.discount.application.strategy;
+
+/**
+ * @author Jiaxu Wei
+ */
+public interface CommerceDiscountApplicationStrategyRegistry {
+
+	public boolean containsKey(String commerceDiscountApplicationStrategyKey);
```
Comment (dantewang) [inline modules/apps/commerce/commerce-discount-api/src/main/java/com/liferay/commerce/discount/application/strategy/CommerceDiscountApplicationStrategyRegistry.java:22]: @W2jx Please review the usages of `containsKey` method. If all the usages are followed by `get`, you can remove `containsKey`.

Just do

```
A a = registry.get(key);

if (a == null) {
    xxx;
}

// use a
```

**Review by dantewang — CHANGES_REQUESTED** [review]

File: modules/apps/commerce/commerce-discount-api/src/main/java/com/liferay/commerce/discount/application/strategy/CommerceDiscountApplicationStrategyRegistry.java (hunk 2 above)
Comment (dantewang) [reply modules/apps/commerce/commerce-discount-api/src/main/java/com/liferay/commerce/discount/application/strategy/CommerceDiscountApplicationStrategyRegistry.java:22]: @W2jx By the way, even if there's usage of `containsKey` without following `get`, you should convert the usages of `containsKey` and `get` which are used together to the form above.

Comment (dantewang) [reply modules/apps/commerce/commerce-discount-api/src/main/java/com/liferay/commerce/discount/application/strategy/CommerceDiscountApplicationStrategyRegistry.java:22]: @W2jx And I think all the `containsKey` usages can actually replaced with `get` + null check.

**Comment by W2jx** [comment]
Hi @dantewang , I have replaced method containsKey and made some sf.

**Comment by W2jx** [comment]
ci:test:relevant

**Comment by W2jx** [comment]
ci:test:sf

**Comment by Kvale1733** [comment]
ci:test:commerce

**Comment by Kvale1733** [comment]
@W2jx No regressions found.

**Comment by W2jx** [comment]
no conflicts(1/11 16:12)

**Comment by dantewang** [comment]
@W2jx The pull looks good to me, but if you want to replace osgi property with interface method, I'd like to see you make this change first, apply, and then convert to service tracker maps. Please rearrange your changes and resend. Thanks!

**Pull request closed without merging**